### PR TITLE
Syscall mock testing

### PR DESCRIPTION
**Pull request by milseman**
Adds mocking and tracing infrastructure, test
infrastructure, and trace-based tests.

TODO: Linux testing.

### CODE REVIEW

File: Sources/SystemInternals/Mocking.swift [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+
+internal let key: pthread_key_t = {
+  var raw = pthread_key_t()
+  func releaseObject(_ raw: UnsafeMutableRawPointer) -> () {
+    Unmanaged<MockingDriver>.fromOpaque(raw).release()
+  }
+  guard 0 == pthread_key_create(&raw, releaseObject) else {
+    fatalError("Unable to create key")
+  }
+  // TODO: All threads are sharing the same object; this is wrong
```
[the remark this answers is not in the document]
Comment (lorentey) [reply Sources/SystemInternals/Mocking.swift:104]: Oh, this looks doable -- the key thing is that you only want to create the key here, but leave its value NULL. The `setspecific` call needs to be moved within `withMockingEnabled` -- the mocking driver must only be set on threads that want to mock, and the wrappers need to bypass all mocking if they see `getspecific` return NULL.

[the remark this answers is not in the document]
Comment (lorentey) [reply Sources/SystemInternals/Mocking.swift:104]: Beware of nested `withMockingEnabled` invocations, it's easy to accidentally overwrite an active pointer & leak memory (`setspecific` doesn't run the destructor on the previous value -- so saving & restoring will work)

[the remark this answers is not in the document]
Comment (milseman) [reply Sources/SystemInternals/Mocking.swift:104]: `withMockingEnabled` does swap out the state, but I think we can simplify it by having it swap out a reference. Good idea, and we should also add a test for that.

**Comment by milseman** [comment]
For some reason Github isn't recognizing my latest pushes. Closing and re-opening

**Pull request closed without merging**